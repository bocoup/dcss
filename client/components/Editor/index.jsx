import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Dropdown, Menu, Segment } from 'semantic-ui-react';
import PropTypes from 'prop-types';
import storage from 'local-storage-fallback';
import EditorMenu from '@components/EditorMenu';
import notify from '@components/Notification';
import ScenarioEditor from '@components/ScenarioEditor';
import ScenarioStatusMenuItem from '@components/EditorMenu/ScenarioStatusMenuItem';
import Scenario from '@components/Scenario';
import Slides from './Slides';
import { getScenario, setScenario } from '@client/actions/scenario';

import './editor.css';

class Editor extends Component {
  constructor(props) {
    super(props);

    this.copyScenario = this.copyScenario.bind(this);
    this.deleteScenario = this.deleteScenario.bind(this);
    this.getAllTabs = this.getAllTabs.bind(this);
    this.getPostSubmitCallback = this.getPostSubmitCallback.bind(this);
    this.getSubmitCallback = this.getSubmitCallback.bind(this);
    this.getTab = this.getTab.bind(this);
    this.onClick = this.onClick.bind(this);
    this.onClickScenarioAction = this.onClickScenarioAction.bind(this);
    this.setActiveView = this.setActiveView.bind(this);
    this.updateScenario = this.updateScenario.bind(this);

    let {
      activeTab,
      activeSlideIndex,
      isCopyScenario,
      isNewScenario,
      match,
      scenarioId
    } = this.props;

    if (!scenarioId) {
      scenarioId = isNewScenario ? 'new' : match.params.id;
    }

    if (isNewScenario) {
      activeTab = 'scenario';
    }

    if (!isCopyScenario && !isNewScenario) {
      this.persistenceKey = `editor/${scenarioId}`;

      let persisted = JSON.parse(storage.getItem(this.persistenceKey));

      if (!persisted) {
        persisted = { activeTab: 'scenario', activeSlideIndex };
        storage.setItem(this.persistenceKey, JSON.stringify(persisted));
      }

      // These have already been declared as let bindings above
      // but we may override those values here, with whatever
      // was persisted for this scenario
      ({ activeTab, activeSlideIndex } = persisted);
    }

    this.state = {
      activeSlideIndex,
      activeTab,
      saving: false,
      scenarioId,
      tabs: null
    };

    if (isCopyScenario) {
      this.copyScenario(scenarioId);
    }

    if (!isNewScenario) {
      this.state.activeTab = 'slides';
      this.props.getScenario(scenarioId);
    }
  }

  componentDidMount() {
    this.setState(state => ({
      tabs: this.getAllTabs(state.scenarioId)
    }));
  }

  onClick(e, { name: activeTab }) {
    this.setState({ activeTab });

    const { activeSlideIndex, scenarioId } = this.state;
    const persisted = JSON.parse(storage.getItem(this.persistenceKey));
    const updated = {
      ...persisted,
      activeSlideIndex,
      activeTab
    };

    storage.setItem(this.persistenceKey, JSON.stringify(updated));

    const activeNonZeroSlideIndex = updated.activeSlideIndex + 1;
    const pathname = `/editor/${scenarioId}/${activeTab}/${activeNonZeroSlideIndex}`;
    this.props.history.push(pathname);
  }

  setActiveView({ activeTab, activeSlideIndex }) {
    const { scenarioId } = this.state;
    storage.setItem(
      this.persistenceKey,
      JSON.stringify({ activeTab, activeSlideIndex })
    );
    const activeNonZeroSlideIndex = activeSlideIndex + 1 || 1;
    const pathname = `/editor/${scenarioId}/${activeTab}/${activeNonZeroSlideIndex}`;
    this.props.history.push(pathname);

    // TODO: verify that this is correct
    this.setState({ activeSlideIndex });
  }

  onClickScenarioAction(event, data) {
    if (data.name === 'save-scenario') {
      this.updateScenario();
    }

    if (data.name === 'save-status') {
      this.updateScenario({ status: data.id });
    }
  }

  async copyScenario(scenarioId) {
    if (!scenarioId) return;

    const { scenario, status } = await (await fetch(
      `/api/scenarios/${scenarioId}/copy`,
      {
        method: 'POST'
      }
    )).json();

    if (status !== 201) {
      notify({ type: 'error', message: 'Error saving copy.' });
      return;
    }

    // Hard refresh to clear all previous state from the editor.
    location.href = `/editor/${scenario.id}`;
  }

  async deleteScenario(scenarioId) {
    const result = await fetch(`/api/scenarios/${scenarioId}`, {
      method: 'DELETE'
    });
    await result.json();

    // Hard redirect to clear all previous state from the editor.
    location.href = '/';
  }

  async updateScenario(updates = {}) {
    if (this.state.saving) {
      return;
    }

    this.setState({ saving: true });

    // NOTE: this is to support saving the whole
    //       scenario when clicking the [save icon]
    //       that's displayed via EditorMenu.
    const {
      author,
      categories,
      consent,
      description,
      finish,
      status,
      title
    } = this.props;

    const data = {
      author,
      categories,
      consent,
      description,
      finish,
      status,
      title
    };

    Object.assign(data, updates);

    const submitCallback = this.getSubmitCallback();
    const response = await (await submitCallback(data)).json();

    this.setState({ saving: false });

    switch (response.status) {
      case 200:
        this.props.setScenario(response.scenario);
        notify({ type: 'success', message: 'Scenario saved' });
        break;
      default:
        if (response.error) {
          notify({ type: 'error', message: response.message });
        }
        break;
    }
  }

  getTab(name, scenarioId) {
    const { activeSlideIndex } = this.state;
    const { setActiveView } = this;

    switch (name) {
      case 'scenario':
        return (
          <ScenarioEditor
            scenarioId={scenarioId}
            submitCB={this.getSubmitCallback()}
            postSubmitCB={this.getPostSubmitCallback()}
          />
        );
      case 'slides':
        return (
          <Slides
            setActiveSlide={activeSlideIndex =>
              setActiveView({
                activeSlideIndex,
                activeTab: 'slides'
              })
            }
            activeSlideIndex={activeSlideIndex}
            scenarioId={scenarioId}
          />
        );
      case 'preview':
        return (
          <Scenario
            setActiveSlide={activeSlideIndex =>
              setActiveView({
                activeSlideIndex,
                activeTab: 'preview'
              })
            }
            activeSlideIndex={activeSlideIndex}
            scenarioId={scenarioId}
          />
        );
      default:
        return null;
    }
  }

  getAllTabs(scenarioId) {
    switch (scenarioId) {
      case 'new':
        return {
          scenario: this.getTab('scenario', 'new')
        };
      default:
        return {
          scenario: this.getTab('scenario', scenarioId),
          slides: this.getTab('slides', scenarioId),
          preview: this.getTab('preview', scenarioId)
        };
    }
  }

  getSubmitCallback() {
    let endpoint, method;
    const { isNewScenario, scenarioId } = this.props;

    if (isNewScenario) {
      endpoint = '/api/scenarios';
      method = 'PUT';
    } else {
      endpoint = `/api/scenarios/${scenarioId}`;
      method = 'POST';
    }
    return scenario => {
      return fetch(endpoint, {
        method,
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(scenario)
      });
    };
  }

  getPostSubmitCallback() {
    const { history, isCopyScenario, isNewScenario } = this.props;
    const { setActiveView } = this;

    if (isCopyScenario || isNewScenario) {
      return scenario => {
        history.push(`/editor/${scenario.id}`);
        this.setState({ scenarioId: scenario.id }, () => {
          setActiveView({ activeTab: 'slides', activeSlideIndex: 0 });
        });
      };
    }

    return null;
  }

  render() {
    const scenarioStatusMenuItem = this.props.status !== undefined && (
      <ScenarioStatusMenuItem
        key="scenario-status-menu-item"
        name="Set scenario status"
        status={this.props.status}
        onClick={this.onClickScenarioAction}
      />
    );
    if (!this.state.tabs) {
      return null;
    }

    const editTabMenu = Object.keys(this.state.tabs).map(tabType => {
      return (
        <Menu.Item
          key={tabType}
          name={tabType}
          active={this.state.activeTab === tabType}
          onClick={this.onClick}
        />
      );
    });

    return (
      <div>
        <Menu attached="top" tabular>
          {editTabMenu}
        </Menu>

        <Segment attached="bottom" className="editor__content-pane">
          {this.state.scenarioId !== 'new' && (
            <EditorMenu
              type="scenario"
              items={{
                save: {
                  onClick: (...args) => {
                    this.onClickScenarioAction(...args);
                  }
                },
                delete: {
                  onConfirm: () => {
                    this.deleteScenario(this.state.scenarioId);
                  }
                },
                right: [scenarioStatusMenuItem]
              }}
            />
          )}

          {this.state.tabs[this.state.activeTab]}
        </Segment>
      </div>
    );
  }
}

// Note: this silences the warning about "text={}" receiving
// an object, instead of a string.
Dropdown.propTypes.text = PropTypes.any;

Editor.propTypes = {
  author: PropTypes.object,
  activeTab: PropTypes.string,
  activeSlideIndex: PropTypes.number,
  scenarioId: PropTypes.node,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired,
  match: PropTypes.shape({
    path: PropTypes.string,
    params: PropTypes.shape({
      id: PropTypes.node,
      activeSlideIndex: PropTypes.node
    }).isRequired
  }).isRequired,
  location: PropTypes.shape({
    state: PropTypes.shape({
      scenarioCopyId: PropTypes.node
    })
  }).isRequired,
  isCopyScenario: PropTypes.bool,
  isNewScenario: PropTypes.bool,
  getScenario: PropTypes.func.isRequired,
  setScenario: PropTypes.func.isRequired,
  title: PropTypes.string,
  description: PropTypes.string,
  categories: PropTypes.array,
  consent: PropTypes.shape({
    id: PropTypes.number,
    prose: PropTypes.string
  }),
  finish: PropTypes.object,
  status: PropTypes.number
};

function mapStateToProps(state) {
  const {
    author,
    categories,
    consent,
    description,
    finish,
    status,
    title
  } = state.scenario;
  return {
    author,
    categories,
    consent,
    description,
    finish,
    status,
    title
  };
}

const mapDispatchToProps = dispatch => ({
  getScenario: id => dispatch(getScenario(id)),
  setScenario: params => dispatch(setScenario(params))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Editor);
