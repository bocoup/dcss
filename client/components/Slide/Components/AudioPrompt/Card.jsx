import React from 'react';
import { Icon } from '@components/UI';
import { name } from './meta';

const Card = () => (
  <React.Fragment>
    <Icon name="microphone" aria-label={name} />
    {name}
  </React.Fragment>
);

export default React.memo(Card);
