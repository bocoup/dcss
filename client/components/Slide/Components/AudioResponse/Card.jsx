import React from 'react';
import { Popup, Icon } from 'semantic-ui-react';
import { name } from './meta';

const AudioResponseCard = () => (
  <Popup
    content="An audio recorder for users to record responses."
    header={name}
    trigger={
      <React.Fragment>
        <Icon name="microphone" aria-label={name} />
        {name}
      </React.Fragment>
    }
  />
);

export default React.memo(AudioResponseCard);
