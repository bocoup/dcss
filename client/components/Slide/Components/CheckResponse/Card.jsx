import React from 'react';
import { Popup, Icon } from 'semantic-ui-react';

const TextResponseCard = () => (
    <Popup
        content="A checkbox response"
        header="Checkbox Response"
        trigger={
            <React.Fragment>
                <Icon
                    name="check square outline"
                    aria-label="Checkbox Response"
                />
                Checkbox Response
            </React.Fragment>
        }
    />
);

export default React.memo(TextResponseCard);
