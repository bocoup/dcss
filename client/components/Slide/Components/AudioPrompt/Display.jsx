import { type } from './meta';
import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Container, Header, Segment } from '@components/UI';
import PromptRequiredLabel from '../PromptRequiredLabel';
import ResponseRecall from '@components/Slide/Components/ResponseRecall/Display';
import AudioRecorder from '@components/Slide/Components/AudioPrompt/AudioRecorder';
import { getResponse } from '@actions/response';
import '@components/Slide/Components/AudioPrompt/AudioPrompt.css';

class Display extends Component {
  constructor(props) {
    super(props);

    const { persisted = { value: '', transcript: '' } } = this.props;

    this.state = {
      isReady: false,
      autostart: false,
      playing: this.isScenarioRun,
      transcript: persisted.transcript,
      type: '',
      value: persisted.value
    };

    this.created_at = new Date().toISOString();
    this.onChange = this.onChange.bind(this);
  }

  get isScenarioRun() {
    return location.pathname.includes('/run/');
  }

  async componentDidMount() {
    if (!this.isScenarioRun) {
      this.setState({
        isReady: true
      });
      return;
    }
    let {
      getResponse,
      onResponseChange,
      persisted = {},
      responseId,
      run
    } = this.props;

    let { name = responseId, transcript = '', value = '' } = persisted;

    let state = {
      ...this.state
    };

    if (!value || !transcript) {
      const previous = await getResponse({
        id: run.id,
        responseId
      });

      if (previous && previous.response) {
        value = previous.response.value;
        transcript = previous.response.transcript;
      }
    }

    if (value) {
      onResponseChange({}, { name, transcript, value, isFulfilled: true });
      state = {
        ...state,
        transcript,
        value
      };
    }

    this.setState({
      ...state,
      isReady: true
    });
  }

  onChange(event, { name, value }) {
    const { created_at } = this;
    const { recallId } = this.props;
    const transcript = '';
    this.props.onResponseChange(event, {
      created_at,
      ended_at: new Date().toISOString(),
      recallId,
      name,
      transcript,
      type,
      value
    });

    this.setState({ transcript, value });
  }

  render() {
    const {
      isReady,
      autostart,
      isRecording,
      playing,
      transcript,
      value
    } = this.state;

    if (!isReady) {
      return null;
    }

    const {
      prompt,
      recallId,
      responseId,
      required,
      run,
    } = this.props;
    const { onChange } = this;
    const isFulfilled = value ? true : false;
    const header = (
      <Fragment>
        {prompt} {required && <PromptRequiredLabel fulfilled={isFulfilled} />}
      </Fragment>
    );

    const willShowAudioRecorder = isFulfilled || autostart;

    return (
      <Fragment>
        <Segment>
          <Header as="h3">{header}</Header>
          {recallId && <ResponseRecall run={run} recallId={recallId} />}
          <AudioRecorder
            autostart={autostart}
            getResponse={this.props.getResponse}
            isEmbeddedInSVG={this.props.isEmbeddedInSVG}
            isRecording={isRecording}
            onChange={onChange}
            prompt={prompt}
            responseId={responseId}
            run={run}
            transcript={transcript}
            value={value}
          />
        </Segment>
      </Fragment>
    );
  }
}

Display.defaultProps = {
  isEmbeddedInSVG: false
};

Display.propTypes = {
  isEmbeddedInSVG: PropTypes.bool,
  getResponse: PropTypes.func,
  isRecording: PropTypes.bool,
  onResponseChange: PropTypes.func,
  persisted: PropTypes.object,
  placeholder: PropTypes.string,
  prompt: PropTypes.string,
  recallId: PropTypes.string,
  required: PropTypes.bool,
  responseId: PropTypes.string,
  run: PropTypes.object,
  type: PropTypes.oneOf([type]).isRequired
};

const mapStateToProps = state => {
  const { run } = state;
  return { run };
};

const mapDispatchToProps = dispatch => ({
  getResponse: params => dispatch(getResponse(params))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Display);
