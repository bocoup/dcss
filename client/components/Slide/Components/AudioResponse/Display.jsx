import { type } from './meta';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  Button,
  Form,
  Header,
  Icon,
  Message,
  Segment
} from 'semantic-ui-react';
import PromptRequiredLabel from '../PromptRequiredLabel';
import ResponseRecall from '@components/Slide/Components/ResponseRecall/Display';
import MicRecorder from 'mic-recorder-to-mp3';
import { detect } from 'detect-browser';
import { connect } from 'react-redux';
import { getResponse } from '@client/actions/response';
import './AudioResponse.css';

const SUPPORTED_BROWSERS = ['chrome', 'firefox'];

class Display extends Component {
  constructor(props) {
    super(props);

    const { persisted = { value: '', transcript: '' } } = this.props;

    this.state = {
      blobURL: '',
      isRecording: false,
      transcript: persisted.transcript,
      type: '',
      value: persisted.value
    };

    this.created_at = new Date().toISOString();
    this.mp3Recorder = new MicRecorder({ bitRate: 128 });

    this.onChange = this.onChange.bind(this);
    this.onFocus = this.onFocus.bind(this);
    this.onStart = this.onStart.bind(this);
    this.onStop = this.onStop.bind(this);

    this.browserSupported = SUPPORTED_BROWSERS.includes(detect().name);
  }

  get isScenarioRun() {
    return location.pathname.includes('/run/');
  }

  async componentDidMount() {
    if (!this.isScenarioRun) {
      return;
    }
    let {
      getResponse,
      onResponseChange,
      persisted = {},
      responseId,
      run
    } = this.props;

    let { name = responseId, transcript = '', value = '' } = persisted;

    if (!value || !transcript) {
      const previous = await getResponse({
        id: run.id,
        responseId
      });

      if (previous && previous.response) {
        value = previous.response.value;
        transcript = previous.response.transcript;
      }
    }

    if (value) {
      onResponseChange({}, { name, transcript, value, isFulfilled: true });
      this.setState({ transcript, value });
    }
  }

  async onStart() {
    await this.mp3Recorder.start();
    this.created_at = new Date().toISOString();
    this.setState({
      isRecording: true
    });
  }

  async onStop() {
    const [buffer, blob] = await (await this.mp3Recorder.stop()).getMp3();
    const blobURL = URL.createObjectURL(blob);

    const file = new File(buffer, 'recording.mp3', {
      type: blob.type,
      lastModified: Date.now()
    });

    const { responseId, responseId: name, run } = this.props;

    let body = new FormData();
    body.append('name', 'audio-response');
    body.append('recording', file);
    body.append('responseId', responseId);

    if (run) {
      body.append('runId', run.id);
    }

    const { s3Location: value } = await (await fetch('/api/media/audio', {
      method: 'POST',
      body
    })).json();

    this.setState(prevState => {
      if (prevState.blobURL) {
        URL.revokeObjectURL(prevState.blobURL);
      }
      return { blobURL, value, isRecording: false };
    });

    const { created_at } = this;
    const { recallId } = this.props;
    const transcript = '';

    this.props.onResponseChange(
      {},
      {
        created_at,
        ended_at: new Date().toISOString(),
        name,
        recallId,
        transcript,
        type,
        value
      }
    );

    this.setState({ transcript, value });
  }

  onFocus() {
    if (!this.created_at) {
      this.created_at = new Date().toISOString();
    }
  }

  onChange(event, { name, value }) {
    const { created_at } = this;
    const { recallId } = this.props;
    const transcript = '';
    this.props.onResponseChange(event, {
      created_at,
      ended_at: new Date().toISOString(),
      recallId,
      name,
      transcript,
      type,
      value
    });

    this.setState({ transcript, value });
  }

  render() {
    const { isRecording, blobURL, transcript, value } = this.state;
    const { prompt, recallId, responseId, required, run } = this.props;
    const { onChange, onFocus } = this;
    const fulfilled = value || blobURL ? true : false;
    const header = this.browserSupported ? (
      required && <PromptRequiredLabel fulfilled={fulfilled} />
    ) : (
      <React.Fragment>
        {prompt} {required && <PromptRequiredLabel fulfilled={fulfilled} />}
      </React.Fragment>
    );

    return this.browserSupported ? (
      <Segment>
        <Header as="h3">{header}</Header>
        {recallId && <ResponseRecall run={run} recallId={recallId} />}
        {!isRecording && (
          <Button basic toggle onClick={this.onStart}>
            <Icon name="circle" aria-label="Record an Audio Response" />
            {prompt}
          </Button>
        )}
        {isRecording && (
          <Button basic negative onClick={this.onStop}>
            <Icon name="stop circle" aria-label="Record an Audio Response" />
            Done
          </Button>
        )}

        {(blobURL || value) && (
          <Message
            content={
              <React.Fragment>
                <audio
                  src={blobURL || `/api/media/${value}`}
                  controls="controls"
                />

                <blockquote className="audiotranscript__blockquote">
                  {transcript ||
                    '(Transcription in progress. This may take a few minutes, depending on the length of your audio recording.)'}
                </blockquote>
              </React.Fragment>
            }
          />
        )}
      </Segment>
    ) : (
      <Segment>
        <Header as="h3">{header}</Header>

        <Form>
          <Form.TextArea
            name={responseId}
            placeholder="..."
            onFocus={onFocus}
            onChange={onChange}
          />
        </Form>
      </Segment>
    );
  }
}

Display.propTypes = {
  getResponse: PropTypes.func,
  isRecording: PropTypes.bool,
  onResponseChange: PropTypes.func,
  persisted: PropTypes.object,
  placeholder: PropTypes.string,
  prompt: PropTypes.string,
  recallId: PropTypes.string,
  required: PropTypes.bool,
  responseId: PropTypes.string,
  run: PropTypes.object,
  type: PropTypes.oneOf([type]).isRequired
};

const mapStateToProps = state => {
  const { run } = state;
  return { run };
};

const mapDispatchToProps = dispatch => ({
  getResponse: params => dispatch(getResponse(params))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Display);
