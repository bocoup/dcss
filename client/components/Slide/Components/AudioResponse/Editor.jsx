import React, { Component } from 'react';
import PropTypes from 'prop-types';
import hash from 'object-hash';
import { Container, Form, Message } from '@components/UI';
import { type } from './meta';
import DataHeader from '@components/Slide/Components/DataHeader';
import ResponseRecall from '@components/Slide/Components/ResponseRecall/Editor';
import './AudioResponse.css';
import '@components/Slide/SlideEditor/SlideEditor.css';

class AudioResponseEditor extends Component {
  constructor(props) {
    super(props);
    const { header = '', prompt = '', recallId = '', responseId } = props.value;

    this.state = {
      header,
      prompt,
      recallId,
      responseId
    };
    this.onChange = this.onChange.bind(this);
    this.onRecallChange = this.onRecallChange.bind(this);
    this.updateState = this.updateState.bind(this);
    this.delayedUpdateState = this.delayedUpdateState.bind(this);
    this.timeout = null;
  }

  componentWillUnmount() {
    clearTimeout(this.timeout);

    const { header, prompt, recallId, responseId } = this.props.value;

    const lastProps = {
      header,
      prompt,
      recallId,
      responseId
    };

    if (hash(this.state) !== hash(lastProps)) {
      this.updateState();
    }
  }

  delayedUpdateState() {
    if (this.timeout) {
      clearTimeout(this.timeout);
    }

    this.timeout = setTimeout(this.updateState, 500);
  }

  updateState() {
    const { header, prompt, recallId, responseId } = this.state;

    this.props.onChange({
      header,
      prompt,
      recallId,
      responseId,
      type
    });
  }

  onChange(event, { name, value }) {
    this.setState({ [name]: value }, this.delayedUpdateState);
  }

  onRecallChange({ recallId }) {
    this.setState({ recallId }, this.updateState);
  }

  render() {
    const { header, prompt, recallId } = this.state;
    const { scenarioId, slideIndex } = this.props;
    const { onChange, onRecallChange, updateState } = this;

    return (
      <Form>
        <Container fluid>
          <ResponseRecall
            isEmbedded={true}
            value={{ recallId }}
            scenarioId={scenarioId}
            slideIndex={slideIndex}
            onChange={onRecallChange}
          />

          <Form.TextArea
            label="Enter text content to display before the audio recording controls:"
            name="prompt"
            value={prompt}
            onChange={onChange}
            onBlur={updateState}
          />

          <DataHeader
            content={header}
            onChange={onChange}
            onBlur={updateState}
          />

          <Message
            icon="warning sign"
            color="orange"
            content="This component will fallback to a text input prompt when the participant's browser or device does not support audio recording."
          />
        </Container>
      </Form>
    );
  }
}

AudioResponseEditor.propTypes = {
  onChange: PropTypes.func.isRequired,
  scenarioId: PropTypes.any,
  slideIndex: PropTypes.any,
  value: PropTypes.shape({
    id: PropTypes.string,
    header: PropTypes.string,
    prompt: PropTypes.string,
    recallId: PropTypes.string,
    required: PropTypes.bool,
    responseId: PropTypes.string,
    type: PropTypes.oneOf([type])
  })
};

export default AudioResponseEditor;
