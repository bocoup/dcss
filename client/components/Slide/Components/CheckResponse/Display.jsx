import { type } from './type';
import React from 'react';
import PropTypes from 'prop-types';
import { Label, Checkbox } from 'semantic-ui-react';

import './CheckResponse.css';

const Display = ({
    prompt,
    checked,
    unchecked,
    responseId,
    onResponseChange
}) => (
    <React.Fragment>
        <Label as="label" className="CheckResponse-label">
            <Checkbox
                name={responseId}
                value={checked}
                label={prompt}
                onChange={(event, values) => {
                    if (values.checked) {
                        values.value = checked;
                    } else {
                        values.value = unchecked;
                    }
                    if (onResponseChange) onResponseChange(event, values);
                }}
                className="CheckResponse-input"
            />
        </Label>
    </React.Fragment>
);

Display.propTypes = {
    prompt: PropTypes.string,
    checked: PropTypes.string,
    unchecked: PropTypes.string,
    responseId: PropTypes.string,
    onResponseChange: PropTypes.func,
    type: PropTypes.oneOf([type]).isRequired
};

export default React.memo(Display);
