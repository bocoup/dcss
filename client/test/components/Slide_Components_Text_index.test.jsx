import React from 'react';
import assert from 'assert';
import { mounter, snapshotter } from '../bootstrap';

import {
  type,
  name,
  description,
} from '@components/Slide/Components/Text/meta';
import Card from '@components/Slide/Components/Text/Card';
import Display from '@components/Slide/Components/Text/Display';
import Editor from '@components/Slide/Components/Text/Editor';
import * as component from '../../components/Slide/Components/Text/index.js';

beforeEach(() => {});

afterEach(() => {});

test('type', () => {
  assert.equal(component.type, type);
});

test('name', () => {
  assert.equal(component.name, name);
});

test('description', () => {
  assert.equal(component.description, description);
});

test('Display', () => {
  assert.equal(component.Display, Display);
});

test('Editor', () => {
  assert.equal(component.Editor, Editor);
});

test('Card', () => {
  assert.equal(component.Card, Card);

  expect(snapshotter(mounter(component.Card))).toMatchSnapshot();
  expect(snapshotter(mounter(Card))).toMatchSnapshot();
  assert.equal(
    snapshotter(mounter(Card)),
    snapshotter(mounter(component.Card))
  );
});

test('Snapshot 1', () => {
  const params = {};

  expect(component.defaultValue(params)).toMatchSnapshot();
});

/*{INJECTION}*/

