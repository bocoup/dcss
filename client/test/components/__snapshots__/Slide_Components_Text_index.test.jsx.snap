// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card 1`] = `"<i aria-label=\\"Rich Text Editor\\" class=\\"edit icon\\"></i>Rich Text Editor"`;

exports[`Card 2`] = `"<i aria-label=\\"Rich Text Editor\\" class=\\"edit icon\\"></i>Rich Text Editor"`;

exports[`Card 3`] = `
"<body>
  <div
    id=\\"root\\"
  />
  <div>
    <i
      aria-label=\\"Rich Text Editor\\"
      class=\\"edit icon\\"
    />
    Rich Text Editor
  </div>
</body>"
`;
